enhancements/machine-config/pre-assigning-nodes-to-machine-config-pools-4-8.md

```diff
@@ -0,0 +1,269 @@
+last-updated: 2021-03-31
+status: provisional
+see-also:
+  - "/openshift/enhancements/pull/717"
+replaces:
+  - 
+superseded-by:
+  - 
+---
+
+# Short-term approach for Pre-assigning Nodes to Machine Config Pools
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [x] Design details are appropriately documented from clear requirements
+- [x] Test plan is defined
+- [ ] Operational readiness criteria is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+In telco bare-metal environments, there is a need to have Nodes be provisioned
+with a known Machine Config Pool.
+
+The traditional flow is for Nodes to come up as workers and be moved to the
+desired pool after provisioning, however this consumes a significant portion of
+the maintenance window in bare-metal environments.
+
+## Motivation
+
+In telco environments, there is a fixed window (typically 4 hours, half of which
+is reserved for rollback) during which new hardware needs to be installed and
+provisioned.
+
+In the case of remote worker nodes, they will typically need to be a part of a
+specific (non-default) MachineConfigPool, and are provisioned with the correct
+pool based on the Ignition file contents.
+
+However, the MCO/MCD normally manages MCP assignment and does so based on Node
+Labels.  This creates a race to add the necessary labels before the MCO/MCD
+moves the Node back to the default config pool.
+
+On bare metal, the cost of losing this race is significant, and as a result we
+could spend half of the 2 hour window just rebooting.
+
+### Goals
+
+- Provide a CR that defines a set of labels that should be applied based on a Node's name
+- Provide the ability to apply these labels *before* the Node object is created
+- Provide the ability to have changes to the CRs be reflected on existing Nodes
+- Meet critical telco use cases by shipping a solution in 4.8
+
+### Non-Goals
+
+- Establish a long-term direction for addressing this within the product.  See
+  [enhancement #717](https://github.com/openshift/enhancements/pull/717) for
+  discussion around an in-product approach
+
+## Proposal
+
+Two new CRs, a controller to watch each, and a webhook.
+
+The first CR allows the admin to define patterns and a list of labels that
+should be applied to any Node with a matching name.
+
+```go
+type LabelsSpec struct {
+	// NodeNamePatterns defines a list of node name regex patterns for which the given labels should be set.
+	// String start and end anchors (^/$) will be added automatically
+	NodeNamePatterns []string `json:"nodeNamePatterns"`
+
+	// Label defines the labels which should be set if one of the node name patterns matches
+	// Format of label must be domain/name=value
+	Labels map[string]string `json:"labels"`
+}
+```
+
+This CR is used by a MutatingAdmissionWebhook when a new Node is created to
```

Comment:
Is the plan to build the "central" part of a cluster, deploy this admission hook, then add the workers that need the special handling? Or does the hook need to be present and work as part of the initial installation?

Reply:
the former: the operator with this webhook will be deployed with OLM / operatorhub after initial cluster installation, before adding workers which need the special handling.

Reply:
OK, it would be good to be explicit about that here in this doc. We've found that adding a description of an end-to-end flow of how someone would use a feature to be a good way of exposing gaps in the implementation plan.

Reply:
Done